docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md

```diff
@@ -0,0 +1,114 @@
+- **Event Subscriber** is an interface which can be accessed publicly by extension developers and get notified when 
+subscribed to a publisher.
+- If the extension uses different file schemes, users have to handle that from extension itself.
+  (ex: expression file scheme)
+
+## Example
+
+### 1. Implementing a Subscriber
+
+We can implement `EventSubscriber` interface as follows.
+
+```java
+@JavaSPIService("org.ballerinalang.langserver.commons.eventsync.spi.EventSubscriber")
+public class CommandRegisterSubscriber implements EventSubscriber {
+    public static final String NAME = "Command register subscriber";
+
+    @Override
+    public EventKind eventKind() {
+        return EventKind.PROJECT_UPDATE;
+    }
+    
+    @Override
+    public void onEvent(ExtendedLanguageClient client, DocumentServiceContext context,
+                        LanguageServerContext languageServerContext) {
+        checkAndRegisterCommands(context);
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+    
+    public static synchronized void checkAndRegisterCommands(DocumentServiceContext context) {
+        LanguageServerContext serverContext = context.languageServercontext();
+        LSClientLogger clientLogger = LSClientLogger.getInstance(serverContext);
+
+        ServerCapabilities serverCapabilities = serverContext.get(ServerCapabilities.class);
+        if (serverCapabilities.getExecuteCommandProvider() == null) {
+            clientLogger.logTrace("Not registering commands: server isn't a execute commands provider");
+            return;
+        }
+
+        if (!isDynamicCommandRegistrationSupported(serverContext)) {
+            clientLogger.logTrace("Not registering commands: client doesn't support dynamic commands registration");
+            return;
+        }
+        context.workspace().waitAndGetPackageCompilation(context.filePath())
+                .ifPresent(pkgCompilation 
+                        -> compileAndRegisterCommands(serverContext, clientLogger, serverCapabilities, pkgCompilation));
+    }
+}
+```
+### 2. Implementing a Publisher
+
+`EventPublishers` can be implemented extending `AbstractEventPublisher`.
+
+```java
+@JavaSPIService("org.ballerinalang.langserver.eventsync.EventPublisher")
+public class ProjectUpdateEventPublisher extends AbstractEventPublisher {
+    public static final String NAME = "Project update event publisher";
+    private CompletableFuture<Boolean> latestScheduled = null;
+    private static final long DIAGNOSTIC_DELAY = 1;
+    
+    @Override
+    public EventKind getKind() {
+        return EventKind.PROJECT_UPDATE;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public void publish(ExtendedLanguageClient client, LanguageServerContext serverContext,
+                        DocumentServiceContext context) {
+        if (latestScheduled != null && !latestScheduled.isDone()) {
```

Comment:
No need to have an implementation here. Just say `// your implementation here` and below this provide a link to an existing implementation.